In [1]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from statsmodels.stats.outliers_influence import variance_inflation_factor
import statsmodels.api as sm
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

cluster0 인 data를 test 와 val와  train으로 나누기
train : 17년도 1분기 ~ 21년도 2분기
val : 21년도 3분기
test : 21년도 4분기 

In [2]:
dataset1=pd.read_csv('cluster_Label1.csv',index_col=0)
dataset1

year  session s_zone  s_zone_code      s_zone_name  total_sale  \
label                                                                   
1      2021        4   골목상권      2111090          강일동주민센터  6164144274   
1      2021        4   골목상권      2111087  고덕중학교(고덕2동주민센터)  2770689276   
1      2021        4   골목상권      2111084         길동신동아아파트   432396902   
1      2021        4   골목상권      2111083        GS강동자이아파트  3938416223   
1      2021        4   골목상권      2111082           중앙보훈병원   413062055   
...     ...      ...    ...          ...              ...         ...   
1      2017        1   골목상권      2110006          부암동주민센터   949311643   
1      2017        1   골목상권      2110004           대신고등학교   925316009   
1      2017        1   골목상권      2110003          세검정초등학교  2420562470   
1      2017        1   골목상권      2110002          독립문역 1번  8522048012   
1      2017        1   골목상권      2110001           이북5도청사  1803437452   

       t_household  income_range     t_spent  office  ...  apt_0.1b_under  \
label                                                 ...                   
1              826             6           0       4  ...               0   
1             1224             6   995226881       1  ...               1   
1              354             7   480653213       0  ...               2   
1             1379             6  1812225212       1  ...              47   
1              486             7   502756843       0  ...              30   
...            ...           ...         ...     ...  ...             ...   
1              363             8   507596200       1  ...               0   
1              258             6   604905596       0  ...              12   
1             1273             7  2635767850       0  ...              70   
1              349             7   516760357       2  ...               8   
1              436             8  1079739057       0  ...              12   

       apt_0.1b  apt_0.2b  apt_0.3b  apt_0.4b  apt_0.5b  apt_0.6b  \
label                                                               
1             0         0         0         0         0         0   
1           125         9         2         1         0         0   
1            78        90       115        30         1         0   
1           347       216       154        42         0         0   
1            64        23        67        41        54        30   
...         ...       ...       ...       ...       ...       ...   
1             0        20         0         1         0         0   
1            16         0         0         0         0         0   
1           374       107        42        50        18         1   
1            36        20         0        27        22         0   
1            47        26       144        26         9        40   

       apt_avg_width  apt_avg_price  total_living  
label                                              
1                  0              0         70137  
1                 45      180259184        192202  
1                 54      218876408        173406  
1                 49      177925349       1028161  
1                 56      183174132        143133  
...              ...            ...           ...  
1                 93      240388889         17924  
1                 50      110733333        353902  
1                 81      194931818        374027  
1                 70      208056122        268151  
1                122      371391608        148356  

[13757 rows x 44 columns]

In [10]:
X_dataset=dataset1.drop(columns=['year','session','total_sale','s_zone','s_zone_code','s_zone_name'])
y_dataset=dataset1['total_sale']
X_test=X_dataset[:686]
X_train=X_dataset[1374:]
X_val=X_dataset[686:1374]
y_val=y_dataset[686:1374]
y_test=y_dataset[:686]
y_train=y_dataset[1374:]


cluster 0 data를 이용한 linear regression

In [11]:
lr = LinearRegression().fit(X_train, y_train)
y_train_hat = lr.predict(X_train)
print(y_train)
print(y_train_hat)
y_val_hat = lr.predict(X_val)
print(y_val)
print(y_val_hat)

label
1    6136905604
1    2893691821
1     476829376
1    3449896079
1     389321713
        ...    
1     949311643
1     925316009
1    2420562470
1    8522048012
1    1803437452
Name: total_sale, Length: 12383, dtype: int64
[7.10701268e+09 2.54513195e+09 3.49147039e+09 ... 5.18377825e+09
 6.44565664e+09 3.07996210e+09]
label
1    5961650254
1    2928283066
1     386352401
1    3979891644
1     374146168
        ...    
1     644459309
1     744236445
1    1730866844
1    7979637934
1    1529373991
Name: total_sale, Length: 688, dtype: int64
[ 7.10423393e+09  2.42173048e+09  3.40889928e+09  4.12728536e+09
  1.27803989e+09  2.00242085e+09  5.78500827e+09  9.26288573e+09
  6.80974684e+09  5.10666134e+09  2.25680637e+09  1.66490738e+09
  4.31596112e+09  8.14076651e+09  7.85988129e+09  1.50430529e+09
  2.66503054e+09  4.91880873e+09  2.37301807e+09  2.77974932e+09
  4.97026099e+08  1.04312018e+09  1.33114288e+10  3.99289470e+09
  1.55884676e+09  4.18363852e+09  2.65059578e+09  6.8536919

In [17]:
print('performance for TRAIN--------')
print('train MAE : ', mean_absolute_error(y_train, y_train_hat))
print('train RMSE : ', mean_squared_error(y_train, y_train_hat)**0.5)
print('train R_square : ', r2_score(y_train, y_train_hat))

print('performance for VAL--------')
print('val MAE : ', mean_absolute_error(y_val, y_val_hat))
print('val RMSE : ', mean_squared_error(y_val, y_val_hat)**0.5)
print('val R_square : ', r2_score(y_val, y_val_hat))

performance for TRAIN--------
train MAE :  1845439749.8038557
train RMSE :  3276462985.3923917
train R_square :  0.40205955718342157
performance for VAL--------
val MAE :  1946664885.2554874
val RMSE :  3893576259.683723
val R_square :  0.2829496626416472


RMSE: 3893576259.683723, r_sqaure: 0.2829496626416472

P-value 값을 확인 하여 다중공산성의심되는 변수 찾아보기

In [12]:
result = sm.OLS(y_train,X_train).fit()
print(result.summary())

                                 OLS Regression Results                                
Dep. Variable:             total_sale   R-squared (uncentered):             -121621.783
Model:                            OLS   Adj. R-squared (uncentered):        -121986.277
Method:                 Least Squares   F-statistic:                             -333.7
Date:                Mon, 23 May 2022   Prob (F-statistic):                        1.00
Time:                        01:04:11   Log-Likelihood:                     -2.8896e+05
No. Observations:               12383   AIC:                                  5.780e+05
Df Residuals:                   12346   BIC:                                  5.783e+05
Df Model:                          37                                                  
Covariance Type:            nonrobust                                                  
                             coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------

다중공산성이 의심되는 p-value 값이 0.05보다 큰 9개의 컬럼('office','office_b','edu_b','traffic_b','t_household_b','apt_99','apt_0.2b','apt_0.6b','apt_avg_price')에 대해 제거를 해보고 다시 regression 시도

In [20]:
X_p_train=X_train.drop(columns=['office','office_b','edu_b','traffic_b','t_household_b','apt_99','apt_0.2b','apt_0.6b','apt_avg_price'])
X_p_val=X_val.drop(columns=['office','office_b','edu_b','traffic_b','t_household_b','apt_99','apt_0.2b','apt_0.6b','apt_avg_price'])

In [21]:
lr = LinearRegression().fit(X_p_train, y_train)
y_p_train_hat = lr.predict(X_p_train)
print(y_train)
print(y_p_train_hat)
y_p_val_hat = lr.predict(X_p_val)
print(y_val)
print(y_p_val_hat)
print('performance for pTRAIN--------')
print('ptrain MAE : ', mean_absolute_error(y_train, y_p_train_hat))
print('ptrain RMSE : ', mean_squared_error(y_train, y_p_train_hat)**0.5)
print('ptrain R_square : ', r2_score(y_train, y_p_train_hat))

print('performance for pVAL--------')
print('val MAE : ', mean_absolute_error(y_val, y_p_val_hat))
print('val RMSE : ', mean_squared_error(y_val, y_p_val_hat)**0.5)
print('val R_square : ', r2_score(y_val, y_p_val_hat))


label
1    6136905604
1    2893691821
1     476829376
1    3449896079
1     389321713
        ...    
1     949311643
1     925316009
1    2420562470
1    8522048012
1    1803437452
Name: total_sale, Length: 12383, dtype: int64
[6.95618013e+09 2.74264689e+09 3.52243901e+09 ... 5.04795435e+09
 6.39289032e+09 3.18134931e+09]
label
1    5961650254
1    2928283066
1     386352401
1    3979891644
1     374146168
        ...    
1     644459309
1     744236445
1    1730866844
1    7979637934
1    1529373991
Name: total_sale, Length: 688, dtype: int64
[ 6.95321945e+09  2.61116655e+09  3.43446211e+09  4.05445987e+09
  1.27195379e+09  2.00435145e+09  5.68316577e+09  9.28577349e+09
  6.66049006e+09  5.14737390e+09  2.21003420e+09  1.62421914e+09
  4.41327789e+09  8.07438950e+09  7.89332969e+09  1.38389702e+09
  2.58867888e+09  4.98977686e+09  2.31850600e+09  2.81137739e+09
  5.46113701e+08  9.74566508e+08  1.32542730e+10  4.02572158e+09
  1.43790550e+09  4.23591537e+09  2.84045031e+09  7.5467657

rmse:  3885966206.0678306, r_square:0.2857498948488948

rmse는 감소했지만 r2의 값은 증가하였으므로 9개의 변수를 지운것이 효과가 있다.

p-value로 다중공산성이 의심되는 변수7개 가 제거된상태에서 vif값을 파악하여 다시한번 다중공산성이 높은 변수를 제거해보도록 하겠다.

In [22]:
vif = pd.DataFrame()
vif["VIF Factor"] = [variance_inflation_factor(X_p_train.values,i) for i in range(X_p_train.shape[1])]
vif["features"]= X_p_train.columns

In [23]:
vif.round(1)

VIF Factor                features
0         11.5             t_household
1         11.7            income_range
2         -0.0                 t_spent
3          1.4                    bank
4          2.0                 medical
5          1.0                     edu
6          1.2                   store
7          2.3                 traffic
8          5.5                  bank_b
9          5.8               medical_b
10         1.6                 store_b
11         1.8              work_10num
12         2.4    work_20_30_40_50_num
13         1.8             work_60_num
14         1.8            work_10num_b
15         4.5  work_20_30_40_50_num_b
16         5.3           work_60_num_b
17        10.0       apt_complexes_num
18        35.9            apt_66_under
19         5.4                  apt_66
20         2.2                 apt_132
21         2.2                 apt_165
22         5.2          apt_0.1b_under
23        19.6                apt_0.1b
24         3.4                apt_0.3b
25         1.9                apt_0.4b
26         1.8                apt_0.5b
27         8.7           apt_avg_width
28         5.8            total_living

vif값이 10이상인 변수(총3개)를 제거

In [24]:
vifX = vif.iloc[vif['VIF Factor'].nlargest(3).index].features.values.tolist()
vifX

['apt_66_under', 'apt_0.1b', 'income_range']

In [25]:
X_vif_train=X_p_train.drop(columns=vifX)
X_vif_val=X_p_val.drop(columns=vifX)

In [27]:
lr = LinearRegression().fit(X_vif_train, y_train)
y_vif_train_hat = lr.predict(X_vif_train)
print(y_train)
print(y_vif_train_hat)
y_vif_val_hat = lr.predict(X_vif_val)
print(y_val)
print(y_vif_val_hat)
print('performance for vifTRAIN--------')
print('train MAE : ', mean_absolute_error(y_train, y_vif_train_hat))
print('train RMSE : ', mean_squared_error(y_train, y_vif_train_hat)**0.5)
print('train R_square : ', r2_score(y_train, y_vif_train_hat))

print('performance for vifVAL--------')
print('val MAE : ', mean_absolute_error(y_val, y_vif_val_hat))
print('val RMSE : ', mean_squared_error(y_val, y_vif_val_hat)**0.5)
print('val R_square : ', r2_score(y_val, y_vif_val_hat))

label
1    6136905604
1    2893691821
1     476829376
1    3449896079
1     389321713
        ...    
1     949311643
1     925316009
1    2420562470
1    8522048012
1    1803437452
Name: total_sale, Length: 12383, dtype: int64
[6.75475918e+09 2.70471140e+09 3.34500723e+09 ... 5.02492778e+09
 6.24457511e+09 3.19624982e+09]
label
1    5961650254
1    2928283066
1     386352401
1    3979891644
1     374146168
        ...    
1     644459309
1     744236445
1    1730866844
1    7979637934
1    1529373991
Name: total_sale, Length: 688, dtype: int64
[ 6.75185594e+09  2.57578145e+09  3.25873687e+09  4.07026585e+09
  1.20344608e+09  2.35732820e+09  5.54537311e+09  9.26407160e+09
  6.58651030e+09  5.34173824e+09  2.14560215e+09  1.72351213e+09
  4.26549086e+09  7.76680075e+09  7.59910562e+09  1.35434326e+09
  2.69268371e+09  5.04524180e+09  2.25656169e+09  2.40707352e+09
  7.91469095e+08  9.98715355e+08  1.34958496e+10  3.67288948e+09
  1.36825884e+09  3.60444207e+09  2.87890604e+09  9.6794179

rmse:3895559410.170271, r_square:0.28221903312503416

vif를 이용하여 변수 3개를  추가 제거해보니 rmse는 커지고 r_sqaure은 작아지므로 성능은 좋아지지 않았다. 따라서 p-value로 9개의 변수를 제거한 데이터를 이용하겠습니다.

pca(0.95정도)를 이용한 feature extraction

In [30]:
scaler = StandardScaler()
scaler.fit(X_p_train)
X_train_scaled = scaler.transform(X_p_train)
X_val_scaled = scaler.transform(X_p_val)
pca = PCA(n_components= 0.95) 
pca.fit(X_train_scaled)
X_train_pca = pca.transform(X_train_scaled)
X_val_pca = pca.transform(X_val_scaled)
print("Original shape: {}".format(str(X_train_scaled.shape)))
print("Reduced shape: {}".format(str(X_train_pca.shape)))

Original shape: (12383, 29)
Reduced shape: (12383, 22)


29=>22 총 7개의 feature가 줄어들었다.

In [31]:
lr = LinearRegression().fit(X_train_pca, y_train)
y_train_pca_hat = lr.predict(X_train_pca)
print(y_train)
print(y_p_train_hat)
y_val_pca_hat = lr.predict(X_val_pca)
print(y_val)
print(y_val_pca_hat)
print('performance for TRAIN--------')
print('train MAE : ', mean_absolute_error(y_train, y_train_pca_hat))
print('train RMSE : ', mean_squared_error(y_train, y_train_pca_hat)**0.5)
print('train R_square : ', r2_score(y_train, y_train_pca_hat))

print('performance for pcaVAL--------')
print('p MAE : ', mean_absolute_error(y_val, y_val_pca_hat))
print('p RMSE : ', mean_squared_error(y_val, y_val_pca_hat)**0.5)
print('p R_square : ', r2_score(y_val, y_val_pca_hat))

label
1    6136905604
1    2893691821
1     476829376
1    3449896079
1     389321713
        ...    
1     949311643
1     925316009
1    2420562470
1    8522048012
1    1803437452
Name: total_sale, Length: 12383, dtype: int64
[6.95618013e+09 2.74264689e+09 3.52243901e+09 ... 5.04795435e+09
 6.39289032e+09 3.18134931e+09]
label
1    5961650254
1    2928283066
1     386352401
1    3979891644
1     374146168
        ...    
1     644459309
1     744236445
1    1730866844
1    7979637934
1    1529373991
Name: total_sale, Length: 688, dtype: int64
[ 6.80914889e+09  2.58809342e+09  3.41536730e+09  4.07673356e+09
  1.32325286e+09  1.98612155e+09  5.75232887e+09  9.51022748e+09
  6.83706800e+09  5.34111969e+09  2.38108229e+09  1.60966983e+09
  4.00494356e+09  7.18724704e+09  8.24965373e+09  1.66275093e+09
  3.03902218e+09  4.82512111e+09  2.33246271e+09  2.32016217e+09
  4.76363136e+08  1.15562892e+09  1.29095862e+10  3.61423174e+09
  1.38388243e+09  3.98213134e+09  2.78786734e+09  1.1919464

rmse:3899153468.2389636, r_square:0.2808939671995526

최종 모델 성능 확인하기 : 다중공산성을 줄이기 위해 vif 와 p-value를 이용하여 변수 총 9개를 제거해보니 성능은 좋아졌으며, 그 후 pca로 feature extraction을 시도해보았지만 성능은 좋아지지 않았다. 따라서 p-value로  총 9개의 컬럼이 지워진 데이터로 test를 진행해보겠다.

In [32]:
X_p_test=X_test.drop(columns=['office','office_b','edu_b','traffic_b','t_household_b','apt_99','apt_0.2b','apt_0.6b','apt_avg_price'])
lr = LinearRegression().fit(X_p_train, y_train)
y_train_hat = lr.predict(X_p_train)
print(y_train)
print(y_train_hat)
y_test_hat = lr.predict(X_p_test)
print(y_test)
print(y_test_hat)
print('performance for TRAIN--------')
print('train MAE : ', mean_absolute_error(y_train, y_train_hat))
print('train RMSE : ', mean_squared_error(y_train, y_train_hat)**0.5)
print('train R_square : ', r2_score(y_train, y_train_hat))

print('performance for test--------')
print('test MAE : ', mean_absolute_error(y_test, y_test_hat))
print('test RMSE : ', mean_squared_error(y_test, y_test_hat)**0.5)
print('test R_square : ', r2_score(y_test, y_test_hat))

label
1    6136905604
1    2893691821
1     476829376
1    3449896079
1     389321713
        ...    
1     949311643
1     925316009
1    2420562470
1    8522048012
1    1803437452
Name: total_sale, Length: 12383, dtype: int64
[6.95618013e+09 2.74264689e+09 3.52243901e+09 ... 5.04795435e+09
 6.39289032e+09 3.18134931e+09]
label
1    6164144274
1    2770689276
1     432396902
1    3938416223
1     413062055
        ...    
1     673924160
1     619901202
1    2596692568
1    9209514193
1    1703154580
Name: total_sale, Length: 686, dtype: int64
[ 6.96111564e+09  2.53431394e+09  3.43745601e+09  4.04251430e+09
  1.26791392e+09  2.04798623e+09  5.72058012e+09  9.35131507e+09
  6.61479024e+09  5.26162978e+09  2.20014576e+09  1.67098234e+09
  3.55928307e+09  7.79696722e+09  7.88638294e+09  1.23732529e+09
  2.65115693e+09  5.10817547e+09  2.30299844e+09  2.85163690e+09
  5.45289268e+08  9.73594912e+08  1.31764598e+10  4.09361303e+09
  1.43073405e+09  4.20255989e+09  2.83969867e+09  7.8153945

cluste1의 데이터를 가지고 모델을 진행 한결과 test RMSE :  3664658386.4731975, test R_square :  0.33699369737036255 임을 알수있습니다.

clustering을 하기전의 데이터 성능과 비교했을 때 rmse값은 감소하였지만, r_sqaure값이 0.51=>0.33로 많이 감소하여 성능이 더 안좋아졌음을 알 수 있다.